Empirical snapshot of flaky tests in Qiskit v2.2.3; feedback welcome

We are software engineering researchers studying flaky tests in quantum software (i.e., tests that nondeterministically pass and fail without code changes). In a recent research project, we performed a large-scale dynamic analysis of flaky tests in Qiskit across multiple historical releases (here is a [link to the preprint ](https://arxiv.org/abs/2512.18088)with details).

As a follow-up snapshot, we ran a point-in-time experiment on Qiskit v2.2.3 (which was the latest release at the time of the experiment) using a modern toolchain (Python 3.13, Rust 1.92) to understand what test flakiness looks like today under modern environments.

For this configuration, we observed:
* 58,380 test cases executed
* 1,835 tests exhibiting both passes and failures
* Approximately 3% observed flakiness

This produced a concrete list of flaky tests when executed repeatedly in a controlled environment.

We are opening this issue purely to ask for feedback, not to report bugs or request fixes: **Would empirical information like this be useful to Qiskit maintainers at all, and if so, in what form?**

For example:
* Is a point-in-time list of empirically flaky tests useful without root-cause analysis?
* Are aggregated summaries preferable to raw test lists?
* Or is this type of data unlikely to be actionable given current CI and development workflows?

This is a one-off research artifact, and we want to avoid creating noise or a maintenance burden. If helpful, we can share the raw list and/or aggregated statistics.

We are mainly trying to understand whether our interpretation of this signal aligns with how the project currently approaches testing and CI, and we would appreciate any perspectives from maintainers.

Thanks for your time, and for your continued work on Qiskit.

Hi @jakelishman ,
We investigated the issue and found that the flakiness stemmed from the multi-worker Tox harness. Parallel execution was introducing instability, causing certain test cases to fail intermittently. Specifically, for `python.qasm2.test_arxiv_examples.TestArxivExamples.test_ripple_carry_adder_1_loads` the failure appears to be related to Tox having trouble loading some libraries: 

```python
Traceback (most recent call last):

      File "/app/qiskit/.tox/py313/lib/python3.13/site-packages/ddt.py", line 221, in wrapper
    return func(self, *args, **kwargs)

      File "/app/qiskit/test/python/qasm2/test_arxiv_examples.py", line 336, in test_ripple_carry_adder
    qc.append(majority(), [cin[0], b[0], a[0]])
              ~~~~~~~~^^

      File "/app/qiskit/test/python/qasm2/__init__.py", line 34, in definer
    gate = qiskit.qasm2.parse._DefinedGate(name, definition.num_qubits, arguments, (), ())
           ^^^^^^^^^^^^^^^^^^

    AttributeError: module 'qiskit.qasm2' has no attribute 'parse'
```
 The fix was straightforward: we switched Tox to using a single worker instead of running tasks in parallel. 

After that, the number of flaky tests in version 2.2.3 dropped to just one, namely `python.circuit.test_ucx_y_z.TestUCRXYZ.test_ucy`, which failed once out of 10,000 times with the following traceback: 
```python
    Traceback (most recent call last):

      File "/app/qiskit/test/python/circuit/test_ucx_y_z.py", line 62, in test_ucy
    self.assertTrue(matrix_equal(unitary_desired, unitary, ignore_phase=True))
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

      File "/usr/local/lib/python3.13/unittest/case.py", line 744, in assertTrue
    raise self.failureException(msg)

    AssertionError: False is not true
```
 
Please take a look and let us know your thoughts.
 
In the meantime, we’re reviewing historical data across older releases. So far, we haven’t seen the same errors in the raw logs for older releases that we saw for version 2.2.3, though it’s possible they were suppressed during execution. @HamidRezaKhoram is currently restarting test harnesses to see if the statistics change. This is taking some time, as the original test servers were recently decommissioned and migrating the harnesses to new hardware has been a challenge.

--- a/test/python/circuit/test_ucx_y_z.py
+++ b/test/python/circuit/test_ucx_y_z.py
@@ -62,7 +62,7 @@ class TestUCRXYZ(QiskitTestCase):
                 # Simulate the decomposed gate
                 unitary = Operator(qc)
                 unitary_desired = _get_ucr_matrix(angles, rot_axis)
-                self.assertTrue(matrix_equal(unitary_desired, unitary, ignore_phase=True))
+                self.assertTrue(matrix_equal(unitary_desired, unitary))
 
 
 def _get_ucr_matrix(angles, rot_axis):

--- a/test/python/circuit/test_ucx_y_z.py
+++ b/test/python/circuit/test_ucx_y_z.py
@@ -24,6 +24,7 @@ from qiskit.compiler import transpile
 from qiskit.circuit.library import UCRXGate, UCRYGate, UCRZGate
 from test import QiskitTestCase  # pylint: disable=wrong-import-order
 
+rng = np.random.default_rng(2026_02_05)
 angles_list = [
     [0],
     [0.4],
@@ -31,9 +32,9 @@ angles_list = [
     [0, 0.8],
     [0, 0, 1, 1],
     [0, 1, 0.5, 1],
-    (2 * np.pi * np.random.rand(2**3)).tolist(),
-    (2 * np.pi * np.random.rand(2**4)).tolist(),
-    (2 * np.pi * np.random.rand(2**5)).tolist(),
+    (2 * np.pi * rng.random(2**3)).tolist(),
+    (2 * np.pi * rng.random(2**4)).tolist(),
+    (2 * np.pi * rng.random(2**5)).tolist(),
 ]
 
 rot_axis_list = ["X", "Y", "Z"]
@@ -55,7 +56,9 @@ class TestUCRXYZ(QiskitTestCase):
                 qc.append(gate, q)
 
                 # Decompose the gate
-                qc = transpile(qc, basis_gates=["u1", "u3", "u2", "cx", "id"])
+                qc = transpile(
+                    qc, basis_gates=["u1", "u3", "u2", "cx", "id"], seed_transpiler=2026_02_05
+                )
                 # Simulate the decomposed gate
                 unitary = Operator(qc)
                 unitary_desired = _get_ucr_matrix(angles, rot_axis)

